### PR TITLE
feat(Theming): Adds Dark, White, and g10 themes to GTC

### LINKED ISSUE #929: Refactor _website-side-nav.scss
Just had the pleasure of digging through this nastiness, we can do better.

### LINKED ISSUE #1028: Dark - light theme configuration guide
Hi, 

This is my very first react website. I have zero experience with any languages and found the this theme + gatsby quite user friendly. 

The issue which I have struggled to find documentation on is how to switch between the dark and light theme without editing each component / shadowing the entire theme. I see an indication of this in the styling section but it does not provide the same newbie clarity other sections provide.

Thanks

### PR DESCRIPTION
**Pull request by sstrubberg**
Problem: A user of Gatsby Theme Carbon has only one theme'ing option out of the box. The homepage is a "dark" theme with the interior pages set to a "gray 10" theme. This leads to users shadowing many components only to change color values to reflect the needs of the given design. Future updates to those shadow'd components are often missed because of their shadow configuration.

Closes #1028 
Closes #929 

Proposed changes:

- Create options in `gatsby-config.js` to allow for modifying the theme of the `homepageTheme` and the `interiorTheme`. The options are `g10`, `white`, and `dark`.
- Touch almost every component to ensure the right tokens are where they need to be.
- All three interior page color options will result in a black-100 PageHeader and PageTabs. Will setup a separate PR for dealing with header customization.

To test:

Follow [the guidance](https://gatsby-theme-carbon-git-fork-sstrubberg-themeing.carbon-design-system.vercel.app/guides/configuration#themeing) and modify the gatsby-config locally to build each variation of the site.

[Sketch file](https://ibm.box.com/shared/static/04439eoqlejcvsf0nmslt5asd5mi77iv.sketch)
[Notifications spec](https://ibm.box.com/s/zcjsg97uhisjvdlt1xf868r9z5xrh61i)

### CODE REVIEW

**Comment by vpicone** [comment]
What do you think about changing the config option API to something like this:

Screen Shot 2020-09-17 at 3 04 42 PM

We can do whatever we want internally and I actually like not doing weird nested queries so keeping it flat here is fine:
Screen Shot 2020-09-17 at 3 04 50 PM

This could let us have other options for the theme in the future without constantly adding top-level options

File: packages/example/src/pages/guides/configuration.mdx [hunk 1]
```diff
@@ -16,6 +16,7 @@ file.
 <AnchorLinks>
   <AnchorLink>Site metadata</AnchorLink>
   <AnchorLink>Manifest</AnchorLink>
+  <AnchorLink>Themeing</AnchorLink>
```
Comment (jnm2377) [inline packages/example/src/pages/guides/configuration.mdx:19]:
```suggestion
  <AnchorLink>Theming</AnchorLink>
```

File: packages/example/src/pages/guides/configuration.mdx [hunk 2]
```diff
@@ -74,6 +75,26 @@ siteMetadata: {
   ],
 ```
 
+## Themeing
```
Comment (jnm2377) [inline packages/example/src/pages/guides/configuration.mdx:78]:
```suggestion
## Theming
```

File: packages/gatsby-theme-carbon/src/components/AnchorLinks/AnchorLinks.module.scss [hunk 3]
```diff
@@ -52,18 +47,18 @@
...
-    &::before {
-      color: $carbon--blue-40;
-    }
-  }
-}
+//     &::before {
+//       color: $carbon--blue-40;
+//     }
+//   }
+// }
```
Comment (jnm2377) [inline packages/gatsby-theme-carbon/src/components/AnchorLinks/AnchorLinks.module.scss:64]: Is this part still WIP or did you mean to delete the comments?

File: packages/gatsby-theme-carbon/src/components/Code/getTheme.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      {
+        types: ['important', 'bold'],
+        style: {
+          fontWeight: 'bold',
+        },
+      },
+      {
+        types: ['important'],
+        style: {
+          color: text01,
```
Comment (jnm2377) [inline packages/gatsby-theme-carbon/src/components/Code/getTheme.js:118]: Why are all the colors `text01`? Don't we want different syntax highlighting?

File: packages/gatsby-theme-carbon/src/templates/Default.js [hunk 5]
```diff
@@ -9,10 +9,19 @@ import EditLink from '../components/EditLink';
...
 import Main from '../components/Main';
+import useMetadata from '../util/hooks/useMetadata';
 
 const Default = ({ pageContext, children, location, Title }) => {
   const { frontmatter = {}, relativePagePath, titleType } = pageContext;
-  const { tabs, title, theme, description, keywords } = frontmatter;
+  const {
+    tabs,
+    title,
+    theme: frontmatterTheme,
```
Comment (jnm2377) [inline packages/gatsby-theme-carbon/src/templates/Default.js:19]: Can you explain where `frontmatterTheme` comes from? Can we set theming in the frontmatter for pages?

**Review by jnm2377 — CHANGES_REQUESTED** [review]
Theming is looking really good! But it looks like we have syntax highlight regression in code snippet, the highlighting is gone. And a few minor bugs: 

1. text in the homepage header seems too dark for the bg. 
Screen Shot 2020-10-01 at 12 57 32 PM

2. no highlighting in codesnippet, but also did we change the design to be white bg?
Screen Shot 2020-10-01 at 12 55 21 PM

File: packages/gatsby-theme-carbon/src/components/Code/getTheme.js (hunk 4 above)
Comment (sstrubberg) [reply packages/gatsby-theme-carbon/src/components/Code/getTheme.js:118]: This is what we need to talk about at our meeting. Basically, the syntax highlighting works great on a dark background, but it falls apart on white or g10 background. Plus, to add to that, [our own guidance](https://www.carbondesignsystem.com/components/code-snippet/usage#formatting) doesn't even suggest syntax highlighting.

File: packages/gatsby-theme-carbon/src/templates/Default.js (hunk 5 above)
Comment (sstrubberg) [reply packages/gatsby-theme-carbon/src/templates/Default.js:19]: This was a recommendation @vpicone gave me. Let me see if I can summarize it. We wanted to need to rename the `theme` prop to `frontmatterTheme` (which is what that line is doing). Later, on line 52, we are assigning it's value || the value from `interiorTheme` to the theme const. This determines the color of the default pages. Vince, can you double check my math here ;)

File: packages/gatsby-theme-carbon/src/components/AnchorLinks/AnchorLinks.module.scss (hunk 3 above)
Comment (sstrubberg) [reply packages/gatsby-theme-carbon/src/components/AnchorLinks/AnchorLinks.module.scss]: Thx!

**Review by jnm2377 — APPROVED** [review]
🎉 LTGM!!!! Great work on this!!!!!! 🎉 💯 🔥 🥇

**Comment by sstrubberg** [comment]
> Theming is looking really good! But it looks like we have syntax highlight regression in code snippet, the highlighting is gone. And a few minor bugs:
> 
> 1. text in the homepage header seems too dark for the bg.
> 
> Screen Shot 2020-10-01 at 12 57 32 PM
> 
> 1. no highlighting in codesnippet, but also did we change the design to be white bg?
> 
> Screen Shot 2020-10-01 at 12 55 21 PM

fixed!

**Comment by jeanservaas** [comment]
One quick comment:

In the Dark theme, the Do / Don't support colors should be using the support tokens.

So in the Gray 100 theme those colors would be:
image

Right now they don't appear to be changing so maybe they're hard coded:
image

**Comment by sstrubberg** [comment]
> One quick comment:
> 
> In the Dark theme, the Do / Don't support colors should be using the support tokens.
> 
> So in the Gray 100 theme those colors would be:
> image
> 
> Right now they don't appear to be changing so maybe they're hard coded:
> image

fixed!

**Review by jeanservaas — COMMENTED** [review]
Looks great!

Couple things:

The notifications you're using on several pages are the low contrast notifications for the White and G10 themes. 

**Minicard page:**
image

**Aside page:**
image

**Inline Notification page**

image



This is the low contrast info notification for the G100 / G90 theme:

image


There isn't a low contrast border under the Black banner in this situation, the low contrast border only appears to separate fields of the same color:

image

On the **Minicard page**, and the **Resource Card page**, the Github icon needs to be g10, right now it's g100:

image

Is this video background (**on Video page**) supposed to be blank?

image



### Image changes

Can I give you a new image here? On the dark theme it looks really busy in part because half of it is the same color as the background?

image

Also these color images on the **Caption** and **Title pages:**

image

**Comment by sstrubberg** [comment]
> Looks great!
> 
> Couple things:
> 
> The notifications you're using on several pages are the low contrast notifications for the White and G10 themes.
> 
> **Minicard page:**
> image
> 
> **Aside page:**
> image
> 
> **Inline Notification page**
> 
> image
> 
> This is the low contrast info notification for the G100 / G90 theme:
> 
> image
> 
> There isn't a low contrast border under the Black banner in this situation, the low contrast border only appears to separate fields of the same color:
> 
> image
> 
> On the **Minicard page**, and the **Resource Card page**, the Github icon needs to be g10, right now it's g100:
> 
> image
> 
> Is this video background (**on Video page**) supposed to be blank?
> 
> image
> 
> ### Image changes
> Can I give you a new image here? On the dark theme it looks really busy in part because half of it is the same color as the background?
> 
> image
> 
> Also these color images on the **Caption** and **Title pages:**
> 
> image

- [ ] Inline notifications (Need specs from Jeannie)
- [ ] Low contrast border on PageHeader (Is there G80 border on the white and G10 theme, but nothing on the Dark theme?)
- [x] Resource and Mini Card Icons
- [ ] Video - (the video provided starts with a black background. We can always swap it out?)
- [ ] ImageCard - (Jeannie is going to get me a new placeholder image that is theme agnostic)
- [ ] Caption & Title pages - (What should I do with those?)

**Comment by jeanservaas** [comment]
@sstrubberg here are a couple of images to replace the things I called out above:

In place of this image (the color spectrum):

image


you can use this image in both light and dark themes:

Image_ai


and in place of this image 

image


you can use this one in both light and dark themes:

dimensional_cyan

File: packages/example/gatsby-config.js [hunk 6]
```diff
@@ -26,6 +26,10 @@ module.exports = {
             'https://github.com/carbon-design-system/gatsby-theme-carbon',
           subDirectory: '/packages/example',
         },
+        theme: {
```
Comment (vpicone) [inline packages/example/gatsby-config.js:29]: just a reminder to take this out before merging

File: packages/example/src/pages/guides/configuration.mdx [hunk 7]
```diff
@@ -108,7 +131,7 @@ plugins: [
       resolve: 'gatsby-theme-carbon',
       options: {
 		// will get added to default [300, 400, 600]
-        additionalFontWeights: ['200', '200i]
+        additionalFontWeights: ['200', '200i']
```
Comment (vpicone) [inline packages/example/src/pages/guides/configuration.mdx:134]: Can we take this out? It's handled by the variable font now, no need to specify weights

**Review by jeanservaas — APPROVED** [review]
This looks amazing. Can you do one small thing for me before you merge, can you change out the Article card image to this new image:

image

can you use this one instead:

article_card_sample

File: packages/gatsby-theme-carbon/src/components/EditLink/EditLink.module.scss [hunk 8]
```diff
@@ -1,22 +1,14 @@
 .link {
   @include carbon--type-style('body-short-01');
   transition: color $transition--base;
-  color: $text-02;
+  color: currentColor;
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/EditLink/EditLink.module.scss:4]: 🔥

File: packages/gatsby-theme-carbon/src/components/Header/Header.module.scss [hunk 9]
```diff
@@ -7,8 +7,8 @@
 }
 
 header.header {
-  background-color: $ui-05;
-  border-bottom: 1px solid $inverse-02;
+  background-color: $carbon--gray-100;
+  // border-bottom: 1px solid $carbon--gray-10;
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/Header/Header.module.scss:11]: Remove?

File: packages/gatsby-theme-carbon/src/components/InlineNotification/InlineNotification.js [hunk 10]
```diff
@@ -1,41 +1,39 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import classnames from 'classnames';
+import cx from 'classnames';
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/InlineNotification/InlineNotification.js:3]: 🔥

File: packages/gatsby-theme-carbon/src/components/InlineNotification/InlineNotification.js [hunk 11]
```diff
@@ -1,41 +1,39 @@
...
+  return (
+    <Row>
+      <Column colLg={8} colMd={6} className={notificationClasses}>
+        <CarbonInlineNotification
+          lowContrast
+          hideCloseButton
+          kind={kind}
+          title=""
+          subtitle={children}
+          className={cx({ [styles.darkMode]: interiorTheme === 'dark' })}
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/InlineNotification/InlineNotification.js:26]: I don't mind this, it's nice if we add more classnames down the line. But just want to make sure you know this works too.

When there's just one class, it's a tad easier to read IMO

```suggestion
          className={interiorTheme === 'dark' && styles.darkMode}
```

File: packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.module.scss [hunk 12]
```diff
@@ -1,20 +1,87 @@
+// White theme
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.module.scss:1]: Great job nuking those nasty overrides in this file. Is it possible to increase the specificity through classes/more specific selectors than using important?

File: packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.js [hunk 13]
```diff
@@ -8,7 +7,7 @@ import LeftNavItem from './LeftNavItem';
 import LeftNavResourceLinks from './ResourceLinks';
 
 import LeftNavWrapper from './LeftNavWrapper';
-import { sideNavDark } from './LeftNav.module.scss';
+import style from './LeftNav.module.scss';
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.js:10]:
```suggestion
import styles from './LeftNav.module.scss';
```

File: packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.js [hunk 14]
```diff
@@ -38,6 +37,13 @@ const LeftNav = (props) => {
     }
   }, [leftNavScrollTop]);
 
+  const getLeftNavClassNames = () => {
+    if (props.theme === 'dark') {
+      return style.sideNavDark;
+    }
+    return style.sideNavWhite;
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/LeftNav/LeftNav.js:44]:
```suggestion
      return styles.sideNavDark;
    }
    return styles.sideNavWhite;
```

File: packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js [hunk 15]
```diff
@@ -1,14 +1,17 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import cx from 'classnames';
-import { pageHeader, withTabs, text } from './PageHeader.module.scss';
+import style from './PageHeader.module.scss';
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js:4]:
```suggestion
import styles from './PageHeader.module.scss';
```

File: packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js [hunk 16]
```diff
@@ -1,14 +1,17 @@
...
 
 const PageHeader = ({ title, tabs = [] }) => (
-  <div className={cx(pageHeader, { [withTabs]: tabs.length })}>
+  <div
+    className={cx(style.pageHeader, {
+      [style.withTabs]: tabs.length,
+    })}>
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js:10]:
```suggestion
    className={cx(styles.pageHeader, {
      [styles.withTabs]: tabs.length,
    })}>
```

File: packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js [hunk 17]
```diff
@@ -1,14 +1,17 @@
...
     <div className="bx--grid">
       <div className="bx--row">
         <div className="bx--col-lg-12">
-          <h1 id="page-title" className={text}>
+          <h1 id="page-title" className={style.text}>
```
Comment (vpicone) [inline packages/gatsby-theme-carbon/src/components/PageHeader/PageHeader.js:14]:
```suggestion
          <h1 id="page-title" className={styles.text}>
```

**Review by vpicone — CHANGES_REQUESTED** [review]
Dude. Amazing work. 🚀 

Some really small things, LeftNav looks way better but maybe we could refine it a bit more by not using `!important`s

**Review by vpicone — APPROVED** [review]

**Pull request merged**